File: 1665-minimum-initial-energy-to-finish-tasks/solution.py
class Solution:
    def minimumEffort(self, tasks: list[list[int]]) -> int:
        """
        To minimize initial energy, we should prioritize tasks where the gap 
        between 'minimum energy to start' and 'actual energy spent' is largest.
        """
        # Sort tasks by (minimum - actual) in descending order.
        tasks.sort(key=lambda x: x[1] - x[0], reverse=True)
        
        total_initial_energy = 0
        current_energy = 0
        
        for actual, minimum in tasks:
            # If current energy is insufficient to start the task,
            # increment the starting energy by the required difference.
            if current_energy < minimum:
                needed = minimum - current_energy
                total_initial_energy += needed
                current_energy += needed
            
            # Deduct the actual energy spent on the task.
            current_energy -= actual
            
        return total_initial_energy

File: 1665-minimum-initial-energy-to-finish-tasks/solution.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    int minimumEffort(vector<vector<int>>& tasks) {
        // Sort tasks based on the difference (minimum - actual) in descending order.
        // This greedy approach ensures we tackle tasks with the largest "surplus"
        // requirements first, utilizing high initial energy effectively.
        sort(tasks.begin(), tasks.end(), [](const vector<int>& a, const vector<int>& b) {
            return (a[1] - a[0]) > (b[1] - b[0]);
        });

        int totalInitialEnergy = 0;
        int currentEnergy = 0;

        for (const auto& task : tasks) {
            int actual = task[0];
            int minimum = task[1];

            // If current energy is less than what's required to start the task,
            // we must increase our initial energy pool.
            if (currentEnergy < minimum) {
                int deficit = minimum - currentEnergy;
                totalInitialEnergy += deficit;
                currentEnergy += deficit;
            }

            // Spend the actual energy required for the task.
            currentEnergy -= actual;
        }

        return totalInitialEnergy;
    }
};